rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+capacity that would work for the majority of workloads therefore we will consider
+the second one.
+
+Pros:
+* Provides sufficient control to benefit from the cache and to maintain memory
+consumption
+* Doesn't introduce yet another *cache capacity* definition that will simplify using the cache
+
+Cons:
+* There is no guarantee that all kernels in the kernel cache are used by all the
+primitives in the primitive cache. It might happen those primitives in the
+primitive cache has unique kernels and the kernel cache may have unique kernels
+too. This means that memory consumption can be higher that it is now
+
+##### Option 2: Provide API to Control Kernel Cache Capacity
+
+This option suggests adding an API to control kernel cache capacity similar to
+the one for the primitive cache.
+
+Pros:
+* Provides more control over the kernel cache and therefore more flexibility to
+the users. Those users who heavily rely on dynamic shapes can balance between
+primitive cache capacity and the kernel cache capacity
+* User can fully control memory consumption
+
+Cons:
+* Introducing yet another *cache capacity* definition that should be specified.
+Also, we will have to explain to the users the difference between kernel cache and primitive cache capacities as well as how they interoperate with each other
+
+###### C API
+
+```c
+dnnl_status_t dnnl_get_kernel_cache_capacity(int *capacity);
+dnnl_status_t dnnl_set_kernel_cache_capacity(int capacity);
+```
+
+###### C++ API
+
+```cpp
+int get_primitive_cache_capacity();
+void set_primitive_cache_capacity(int capacity);
+```
+
+###### Environment Variable
+
+`ONEDNN_PRIMITIVE_CACHE_CAPACITY=<number>`, where `number` is the kernel cache
+capacity. The default value is 1024. When `number` is 0 the kernel cache is
+disabled.
+
+###### Build Time Control
+
+The kernel cache can be enabled or disabled at the build time using a CMake option
+`ONEDNN_ENABLE_KERNEL_CACHE`.
+
+##### Options Comparison
+
+Both options assume that there will be a reasonable default capacity for the
+kernel cache. Also, both options allow users to control the caches capacity
+therefore, the only difference between the options is in the flexibility.
+The second option is in fact the first one with advanced controlling
+capabilities.
+
+Given that the first option will address the problem with dynamic shapes
+related creation overhead the proposal is to start with the it and add the
+API from the option two if we see any need in it.
+
+### Option 2: Extending Runtime Dimensions Support
+
+This option suggests extending the existing runtime dimensions support for
+relevant primitives.
+
+Pros:
+* Using the existing mechanism, the primitive cache will automatically work
+for primitives created with runtime dimensions
+* Unlike the kernel cache, with this approach the primitive and primitive descriptor
+is not re-created therefore there is no related overhead
+
+Cons:
+* The users have to modify their code to enable runtime dimensions
```

Comment:
Why can we make runtime dimensions default on in oneDNN?

Reply:
Because it's not possible to enable the feature for all primitives and all dimensions. I expanded in this [here](https://github.com/oneapi-src/oneDNN/pull/1637#discussion_r1181828955).